### PR TITLE
Fix the issue that pyspark3 not supported by Livy 0.4

### PR DESCRIPTION
**Pull request by taoliseki**
Livy 0.4 onwards removed the "pyspark3" as the session kind, so sparkmagic receives a 400 response when we create a pyspark3 notebook on Jupyter. So pyspark3 kernel is not working due to this Livy breaking change.

In this PR, we are specifying "spark.yarn.appMasterEnv.PYSPARK_PYTHON" to "python" or "python3" to kick off different python versions, which is the new approach to differentiate pyspark/pyspark3 sessions. This approach works with the old livy version (e.g. 0.3) as well as 0.4 onwards. Also "spark.yarn.appMasterEnv.PYSPARK_PYTHON"  is supported both spark 1.6 and spark 2.

### CODE REVIEW

File: sparkmagic/sparkmagic/utils/configuration.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ def override_all(obj):
 
 # Helpers
 
-def get_livy_kind(language):
+def get_session_kind(language):
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/utils/configuration.py]: Can we please revert the change in name? Was it done for any particular reason?

Comment (taoliseki) [reply sparkmagic/sparkmagic/utils/configuration.py]: Yes there is a reason for that change. 

Basically Livy 0.4 onwards has removed "pyspark3" as session kind, which is a breaking change. For sparkmagic, we switch to use the pyspark setting "spark.pyspark.python" to specify python2/python3 instead of the old approach by using the Livy session kind.

As you can see from my change, we are sticking with pyspark/pyspark3 as the session kind definition for sparkmagic, and then translate that to the correct config settings which are sent to Livy when creating a session. I am renaming the method name because this method is now using the new session definition (for sparkmagic), as opposed to livy session definition (which does not have pyspark3 now).

Does it make sense?

File: sparkmagic/sparkmagic/livyclientlib/livysession.py [hunk 2]
```diff
@@ -110,11 +110,22 @@ def __str__(self):
...
 
+    def _translate_to_livy_kind(self, properties, kind):
+        # Livy does not support "pyspark3" as the kind from 0.4 onwards and only "pyspark" is valid.
+        # That's why we are leveraging PYSPARK_PYTHON_PARAM here to specifyf different python versions.
+        if kind == constants.SESSION_KIND_PYSPARK:
+            python_properties = {constants.PYSPARK_PYTHON_PARAM: constants.LANG_PYTHON}
+            properties[constants.LIVY_CONF_PARAM] = python_properties
+        elif kind == constants.SESSION_KIND_PYSPARK3:
+            properties[constants.LIVY_KIND_PARAM] = constants.SESSION_KIND_PYSPARK
+            python_properties = {constants.PYSPARK_PYTHON_PARAM: constants.LANG_PYTHON3}
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/livyclientlib/livysession.py]: The python3 path should be configurable, and not a constant.

Comment (mpekalski) [reply sparkmagic/sparkmagic/livyclientlib/livysession.py]: Also this kind of replacement of properties breaks configuration JSON, e.g. below is before and after. Basically it lost 'spark.jars.packages' I had passed over in the conf. 


before 
```
 properties: {'driverMemory': '1G', 'executorCores': 4, 'kind': 'pyspark3', 'conf': {'spark.pyspark.python': 'python3', 'spark.jars.packages': 'com.databricks:spark-avro_2.11:4.0.0'}, 'heartbeatTimeoutInSecond': 120}
```
after
```
 properties: {'driverMemory': '1G', 'executorCores': 4, 'kind': 'pyspark', 'conf': {'spark.pyspark.python': 'python3'}, 'heartbeatTimeoutInSecond': 120}
```

File: sparkmagic/sparkmagic/utils/configuration.py (hunk 1 above)
Comment (aggFTW) [reply sparkmagic/sparkmagic/utils/configuration.py]: This change is not merged.

Comment (chuyqa) [reply sparkmagic/sparkmagic/utils/configuration.py]: 0.4 Release Still has `pyspark3` https://github.com/apache/incubator-livy/blob/branch-0.4/core/src/main/scala/org/apache/livy/sessions/Kind.scala

0.5 Removes it https://github.com/apache/incubator-livy/blob/branch-0.5/core/src/main/scala/org/apache/livy/sessions/Kind.scala#L40

**Comment by gm-spacagna** [comment]
Hi I have tried to to fix this compatibility issue in a different way in this PR: https://github.com/jupyter-incubator/sparkmagic/pull/507
Would you mind having a look?

**Comment by ranjitiyer** [comment]
@gm-spacagna I have a PR that's failing build https://github.com/jupyter-incubator/sparkmagic/pull/462. Any tips on how to resolve it?

File: sparkmagic/sparkmagic/utils/configuration.py (hunk 1 above)
Comment (kanastasiou4signifyd) [reply sparkmagic/sparkmagic/utils/configuration.py]: Has anyone tested this on livy 0.5?

**Comment by foscraig** [comment]
@gm-spacagna @aggFTW Can you please answer the question? Would like someone to look at that PRs and it's been almost a year, so want to get Travis CI working.

**Comment by foscraig** [comment]
Also tagging @apetresc in case he knows the answer to @ranjitiyer's question. Thanks!

**Comment by itamarst** [comment]
This is superseded by #540, which I hope to get merged soon. Please try it!
